pkg/oidcclient/login.go

```diff
@@ -515,12 +524,14 @@ func (h *handlerState) cliBasedAuth(authorizeOptions *[]oauth2.AuthCodeOption) (
 }
 
 // Prompt for the user's username and password, or read them from env vars if they are available.
-func (h *handlerState) getUsernameAndPassword() (string, string, error) {
+func (h *handlerState) getUsernameAndPassword(out io.Writer) (string, string, error) {
 	var err error
 
+	_, _ = fmt.Fprintf(out, "Log in to %s:\n", h.upstreamIdentityProviderName)
```

Comment:
When `h.upstreamIdentityProviderName` is blank, we should probably skip printing the new login banner. This would be for the pre-multiple IDP's backwards compatibility mode where that can be blank. Note that the tanzu CLI's kubeconfig generation leave this blank, if I recall correctly, so we would not want to print a broken looking banner in that case.

Reply:
Good point.

The original Tracker story for this indicated proactively checking the Supervisor, which could imply all sort of fast-fail scenarios. Examples:

(A) an IDP name is given, but that IDP is not listed on the Federation domain (or does not exist)
(B) no name is given, but the federation domain has multiple IDPs with `cli_password` capability
(C) no name is given, and only one IDP is listed on the federation domain, so we show that IDP name to the user (this isn't a failure)

Reply:
For the purposes of this PR, where we are not making extra calls to the Supervisor to do extra IDP discovery, it seems okay to just skip the banner when the IDP name is blank.

In the current implementation:

- (A) this would be a login failure with a good error message 
- (B) if the FederationDomain has `spec.identityProviders` defined, then the client needs to send the IDP name, so this would end up being similar to (A) in that it would be a login failure with a good error message
- (C) this would be the same as (B), and would be an error

These all seem fine.

When the IDP name is blank, and when the server's FederationDomain has no `spec.identityProviders`, then the server is in the backwards compatibility mode (there must be one and only one `*IdentityProvider` resource) and the login will succeed.

Reply:
For anyone else following this thread, note that the IDP name is not usually blank. The open source `pinniped get kubeconfig` command has been including the IDP name in the resulting kubeconfig for years. So this is an edge case for open source Pinniped (e.g. a user manually edits the kubeconfig), but it's still worth handling since we think the tanzu CLI is going to cause this case. Skipping the banner would preserve the pre-existing behavior for the tanzu CLI.

Reply:
You're right about (C). I said "one IDP is listed on the federation domain" but meant "one IDP is defined in the Supervisor's namespace", which would imply backwards-compatibility mode.